lib/chef/resource/homebrew_install.rb

```diff
@@ -0,0 +1,148 @@
+      ```ruby
+      homebrew_install 'Install Homebrew and xcode command line tools if necessary' do
+        xcode_tools_url 'https://somewhere.something.com/downloads/command_line_tools.dmg'
+        xcode_tools_pkg_name 'Command Line Tools.pkg'
+        user 'someuser'
+        action :install
+      end
+      ```
+      DOC
+
+      property :xcode_tools_url, String,
+        description: "A url pointing to a 'Command Line Tools for Xcode' dmg file"
+
+      property :xcode_tools_pkg_name, String,
+        description: "The name of the pkg inside the dmg located at the xcode_tools_url"
+
+      property :brew_source_url, String,
+        description: "A url pointing to a Homebrew installer",
+        default: "https://codeload.github.com/Homebrew/brew/zip/master"
+
+      property :user, String,
+        description: "The user to install Homebrew as. Note: Homebrew cannot be installed as root.",
+        required: true
+
+      action :install do
+        # Avoid all the work in the below resources if homebrew is already installed
+        return if ::File.exist?("/usr/local/bin/brew")
+
+        # check if 'user' is root and raise an exception if so
+        if Etc.getpwnam(new_resource.user).uid == 0
+          msg = "You are attempting to install Homebrew as Root. This is not permitted by Homebrew. Please run this as a standard user with admin rights"
+          raise Chef::Exceptions::InsufficientPermissions, msg
+        end
+
+        USER_HOME = Dir.home(new_resource.user).freeze
+        HOMEBREW_CACHE = "#{USER_HOME}/Library/Caches/Homebrew".freeze
+
+        # Creating the basic directory structure needed for Homebrew
+        directories = ["bin", "etc", "include", "lib", "sbin", "share", "var", "opt",
+                        "share/zsh", "share/zsh/site-functions",
+                        "var/homebrew", "var/homebrew/linked",
+                        "Cellar", "Caskroom", "Homebrew", "Frameworks"
+                      ].freeze
+        directories.each do |dir|
+          directory "/usr/local/#{dir}" do
+            mode "0755"
+            owner new_resource.user
+            group "admin"
+            action :create
+          end
+        end
+
+        user_directories = ["#{USER_HOME}", "#{USER_HOME}/Library",
+                            "#{USER_HOME}/Library/Caches", "#{USER_HOME}/Library/Caches/Homebrew"
+        ]
+        user_directories.each do |dir|
+          directory "#{dir}" do
+            mode "0755"
+            owner new_resource.user
+            group "admin"
+            action :create
+          end
+        end
+
+        if new_resource.xcode_tools_url
+          dmg_package new_resource.xcode_tools_pkg_name do
+            source new_resource.xcode_tools_url
+            type "pkg"
+          end
+        else
+          build_essential "install Command Line Tools for Xcode" do
+            action :upgrade
```

Comment:
Does this need to be an upgrade. That's a very slow code path and maybe not what users are expecting vs. the install

Reply:
It does not. I will make it an Install